File: C++/leetcode203.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
struct ListNode {
      int val;
      ListNode *next;
      ListNode() : val(0), next(nullptr) {}
      ListNode(int x) : val(x), next(nullptr) {}
      ListNode(int x, ListNode *next) : val(x), next(next) {}
  };



class Solution{
public:
    ListNode *removeElements(ListNode *head, int val)
    {
        struct ListNode* dummpny = new ListNode(0,head);
        struct ListNode* p = head;
        while(p->next!=NULL){
            if(p->next->val==val){
                struct ListNode* temp = p->next;
                p->next=p->next->next;
                delete temp;
            }else{
                p = p->next;
            }
        }
        return dummpny->next;
    }
};

File: python/leetcode203.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        # 可能会有删除头结点的情况
        # 添加一个虚拟头节点，对头节点的删除操作与其他节点一样
        pre_node = ListNode(next=head)
        node = pre_node
        while node.next:
            if node.next.val == val:
                node.next = node.next.next
            else:
                node = node.next

        return pre_node.next
